torchrecipes/paved_path/charnn/combined_module.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+CombinedModule includes model and its corresponding transform.
+It's mainly used for inference from raw inputs, which will be converted
+to tensors by transform then pass to model.
+"""
+from typing import Optional
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+from torch.nn import functional as F
+
+
+class CombinedModule(nn.Module):
+    device: str
+
+    def __init__(self, transform: nn.Module, model: nn.Module) -> None:
+        super().__init__()
+        self.transform = transform
+        self.model = model
+        self.device = ""
+
+    def forward(self, text: str) -> str:
+        tokens = self.transform(text)
+        tokens = tokens.unsqueeze(0).to(self.device)
+        generated_ids = self.generate(tokens).squeeze()
+        return self.transform.decode(generated_ids)
+
+    @torch.jit.export
+    def set_device(self, device: str) -> None:
+        self.device = device
+
+    def top_k_logits(self, logits: Tensor, k: int) -> Tensor:
+        v, ix = torch.topk(logits, k)
+        out = logits.clone()
+        out[out < v[:, [-1]]] = -float("Inf")
+        return out
+
+    def generate(
+        self,
+        x: Tensor,
+        steps: int = 100,
+        temperature: float = 1.0,
+        sample: bool = True,
+        top_k: Optional[int] = 10,
+    ) -> Tensor:
+        """
+        take a conditioning sequence of indices in x (of shape (b,t)) and predict the next token in
+        the sequence, feeding the predictions back into the model each time. Clearly the sampling
+        has quadratic complexity unlike an RNN that is only linear, and has a finite context window
+        of block_size, unlike an RNN that has an infinite context window.
```

Comment:
this is `sample()` function right? we should get rid of it? also, wonder if we have some link to the theory of this logic

Reply:
Right, I think it's originally copied from here: https://github.com/karpathy/minGPT/blob/7218bcfa527c65f164de791099de715b81a95106/mingpt/model.py#L283

I can't find the link to the theory of this logic. As we will publish the tutorial with torchvision example, maybe leave it as it is?

Reply:
do we plan to remove "sample()"?

Reply:
yes, resolved in the latest version in Diffs